src/main/java/org/truffleruby/core/module/ModuleNodes.java

```diff
@@ -957,10 +960,27 @@ public Object setConstant(DynamicObject module, String name, Object value) {
                 throw new RaiseException(coreExceptions().nameError(StringUtils.format("wrong constant name %s", name), module, name, this));
             }
 
-            Layouts.MODULE.getFields(module).setConstant(getContext(), this, name, value);
+            final boolean redefined = Layouts.MODULE.getFields(module).setConstant(getContext(), this, name, value);
+            if (redefined) {
+                warnAlreadyInitializedConstant(module, name);
+            }
             return value;
         }
 
+        @TruffleBoundary
+        private void warnAlreadyInitializedConstant(DynamicObject module, String name) {
+            if (warnNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                warnNode = insert(new WarnNode());
+            }
+            final String moduleName = Layouts.MODULE.getFields(module).hasPartialName() ? Layouts.MODULE.getFields(module).getName() + "::" : "";
+            if (getSourceSection() != null) {
+                warnNode.warn(getSourceSection().getSource().getName(), ":", Integer.toString(getSourceSection().getStartLine()), ": warning: already initialized constant ", moduleName, name);
+            } else {
+                warnNode.warn("warning: already initialized constant ", moduleName, name);
+            }
+        }
```

Comment:
This logic should be shared with WriteConstantNode.

Reply:
What's the preferred way to share this logic between the two?

Reply:
If it's simple enough, a static method, otherwise a Node.

Reply:
Fixed.